### PR TITLE
Add neural estimators in JAX

### PR DESCRIPTION
**Pull request by pawel-czyz**
Hi @grfrederic, I added the code we developed yesterday after minor refactoring.

I also include an attempt to make the losses more scalable (i.e., memory increasing linearly rather than quadratically).

Note that it's not tested and we still need the "high-level" estimators, matching our interfaces.

### CODE REVIEW

File: src/bmi/estimators/neural/_interfaces.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from typing import Callable
+
+import jax.numpy as jnp
+
+# A single point, i.e., a vector (dim_x,)
+Point = jnp.ndarray
+
+# A collection of points, i.e., a vector (batch_size, dim_x)
+BatchedPoints = jnp.ndarray
```
Comment (grfrederic) [inline src/bmi/estimators/neural/_interfaces.py:9]: i feel like this is a bit much

**Comment by grfrederic** [comment]
I added an initial implementation of 3 neural estimators (_not_ MINE). In the current configuration they take about 20 seconds to run on my laptop CPU. I had some trouble with the BatchedPoints typing, maybe you could take a look when you have a moment.

i also think it would be nice to add two things:
1. Add an option to include a train/valid split in the estimators. Importantly we want to run the linear memory version of the estimator for evaluation.
2. In cases we don't make this split, I'm fine with returning the last value, as long as we add some simple convergence check. For example we can see that after we hit highest MI (lowest loss), we don't fall more than 10% lower in all remaining epochs. I would actually consider returning the mean MI over all these epochs as the estimate.

**Comment by grfrederic** [comment]
oh also we should hide the `NeuralEstimator` class from the api, that's my mistake :)

**Comment by grfrederic** [comment]
i also noticed that linear versions of NWJ and DV have sth broken with shapes (i initially accidentally used them for training...) and got jax complaining about shapes

**Comment by grfrederic** [comment]
i added the following changes:
1. train/test split for neural estimators using the safe (slow but linear) backend. the split is optional: **if it is _not_ used, the final estimate is calculated over the whole dataset using the safe backend.** if it is used, the final estimate is the highest estimate from all test ones.
2. i added some simple diagnostics for convergence. maybe they should happen in the _nn module though?
3. optional verbose prints, i think they can be pretty useful for debugging these particular estimators :)

**Comment by pawel-czyz** [comment]
Thanks! Regarding the issue:

> I also noticed that linear versions of NWJ and DV have sth broken with shapes (i initially accidentally used them for training...) and got jax complaining about shapes

it was my silly mistake with vmap. In b426e918d3fa510ea7c3e26cca35c734300769e2 I added a fix and the simplest unit test comparing both implementations on small batches of random points.

File: src/bmi/estimators/neural/_interfaces.py (hunk 1 above)
Comment (pawel-czyz) [reply src/bmi/estimators/neural/_interfaces.py:9]: I personally like it, as we aren't currently using [jaxtyping](https://github.com/google/jaxtyping) and without the annotations I'll soon forget what shapes they are supposed to be. 

Does it cause problems with Pyright? Pytype checks work here just fine.

**Comment by pawel-czyz** [comment]
In b0be261c60214165a8a97a465a65a7eccf3f850a I tried to add a unit test, similar to the tests of other estimators (we take a simple distribution and then see how it works). The number of points is probably too small, but it resulted in the following message:

```
WARNING! Test MI reached 0.168, but then dropped to {test_check[1]:.3f}. This might indicate overfitting.
WARNING! Test MI was still increasing when training stopped. This might mean the network has not fully converged.
```

**Comment by pawel-czyz** [comment]
> oh also we should hide the `NeuralEstimator` class from the api, that's my mistake :)

Done! :slightly_smiling_face:

**Comment by grfrederic** [comment]
> In [b0be261](https://github.com/pawel-czyz/bmi/commit/b0be261c60214165a8a97a465a65a7eccf3f850a) I tried to add a unit test, similar to the tests of other estimators (we take a simple distribution and then see how it works). The number of points is probably too small, but it resulted in the following message:
> 
> ```
> WARNING! Test MI reached 0.168, but then dropped to {test_check[1]:.3f}. This might indicate overfitting.
> WARNING! Test MI was still increasing when training stopped. This might mean the network has not fully converged.
> ```

I fixed the formatting. I also changed the printing so that if we detect overfitting (first message), we don't report possible undertraining (MI increasing in last two epochs = the second message).

File: src/bmi/estimators/neural/_interfaces.py (hunk 1 above)
Comment (grfrederic) [reply src/bmi/estimators/neural/_interfaces.py:9]: nope, no problems. i guess lets leave it. doesn't jax allow for annotating tensor dimensions though?

Comment (pawel-czyz) [reply src/bmi/estimators/neural/_interfaces.py:9]: Package jaxtyping allows, but I don't think we have configured it for this project... (sadly)

**Review by grfrederic — APPROVED** [review]
awesome! go ahead, you push the button ;)
